iroh/src/commands.rs

```diff
@@ -0,0 +1,263 @@
+                    }
+                }
+            }
+            Commands::GetTicket { out, ticket } => {
+                let get = self::get::GetInteractive::Ticket {
+                    ticket,
+                    keylog: self.keylog,
+                    derp_map: config.derp_map(),
+                };
+                tokio::select! {
+                    biased;
+                    res = get.get_interactive(out) => res,
+                    _ = tokio::signal::ctrl_c() => {
+                        println!("Ending transfer early...");
+                        Ok(())
+                    }
+                }
+            }
+            Commands::Provide {
+                path,
+                addr,
+                rpc_port,
+            } => self::provide::run(config, rt, path, addr, rpc_port, self.keylog).await,
+            Commands::List(cmd) => cmd.run().await,
+            Commands::Validate { rpc_port } => self::validate::run(rpc_port).await,
+            Commands::Shutdown { force, rpc_port } => {
+                let client = make_rpc_client(rpc_port).await?;
+                client.rpc(ShutdownRequest { force }).await?;
+                Ok(())
+            }
+            Commands::Id { rpc_port } => {
+                let client = make_rpc_client(rpc_port).await?;
+                let response = client.rpc(IdRequest).await?;
+
+                println!("Listening address: {:#?}", response.listen_addrs);
+                println!("PeerID: {}", response.peer_id);
+                Ok(())
+            }
+            Commands::Add { path, rpc_port } => self::add::run(path, rpc_port).await,
+            Commands::Addresses { rpc_port } => {
+                let client = make_rpc_client(rpc_port).await?;
+                let response = client.rpc(AddrsRequest).await?;
+                println!("Listening addresses: {:?}", response.addrs);
+                Ok(())
+            }
+            Commands::Doctor { command } => self::doctor::run(command, config).await,
+        }
+    }
+}
+#[derive(Subcommand, Debug, Clone)]
+#[allow(clippy::large_enum_variant)]
+pub enum Commands {
+    /// Diagnostic commands for the derp relay protocol.
+    Doctor {
+        /// Commands for doctor - defined in the mod
+        #[clap(subcommand)]
+        command: self::doctor::Commands,
+    },
+
+    /// Serve data from the given path.
+    ///
+    /// If PATH is a folder all files in that folder will be served.  If no PATH is
+    /// specified reads from STDIN.
+    Provide {
+        /// Path to initial file or directory to provide
+        path: Option<PathBuf>,
+        #[clap(long, short)]
+        /// Listening address to bind to
+        #[clap(long, short, default_value_t = SocketAddr::from(crate::node::DEFAULT_BIND_ADDR))]
+        addr: SocketAddr,
+        /// RPC port, set to "disabled" to disable RPC
+        #[clap(long, default_value_t = ProviderRpcPort::Enabled(DEFAULT_RPC_PORT))]
+        rpc_port: ProviderRpcPort,
+    },
```

Comment:
I'd vote to move this into `commands/provide.rs`, the current contents of `commands/provide.rs` into `src/provide.rs`, and call a public function from `src/provide.rs` within `commands::provide::run`

Reply:
See comment above, the public api for providing is currently the `Node` builder.